cmd/backup-agent/main.go

```diff
@@ -0,0 +1,69 @@
+package main
+
+import (
+	"fmt"
+	"os"
+
+	flag "github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	etcdv1alpha1 "github.com/improbable-eng/etcd-cluster-operator/api/v1alpha1"
+	"github.com/improbable-eng/etcd-cluster-operator/controllers"
+	"github.com/improbable-eng/etcd-cluster-operator/version"
+)
+
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+func init() {
+	_ = clientgoscheme.AddToScheme(scheme)
+
+	_ = etcdv1alpha1.AddToScheme(scheme)
+	// +kubebuilder:scaffold:scheme
+}
+
+func main() {
+	var (
+		printVersion  bool
+		backupTempDir string
+	)
+
+	flag.BoolVar(&printVersion, "version", false, "Print the version to stdout and exit")
+	flag.StringVar(&backupTempDir, "backup-tmp-dir", os.TempDir(), "The directory to temporarily place backups before they are uploaded to their destination.")
+	flag.Parse()
+
+	if printVersion {
+		fmt.Println(version.Version)
+		os.Exit(0)
+	}
+	ctrl.SetLogger(zap.Logger(true))
+
+	setupLog.Info("Starting backup-agent", "version", version.Version)
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme: scheme,
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to start manager")
+		os.Exit(1)
+	}
+
+	if err = (&controllers.EtcdBackupReconciler{
+		Client:  mgr.GetClient(),
+		Log:     ctrl.Log.WithName("controllers").WithName("EtcdBackup"),
+		TempDir: backupTempDir,
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "EtcdBackup")
+		os.Exit(1)
+	}
+
+	setupLog.Info("starting manager")
+	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
+		setupLog.Error(err, "problem running manager")
+		os.Exit(1)
+	}
```

Comment:
What do we need a controller and reconciler for?

The backup-agent should just receive its information through flags (the urls of the cluster endpoints, the URL of the proxy, the cluster identifier to use when uploading the backup to the proxy, etc.) It shouldn't need to speak to the Kubernetes API for anything.

Reply:
Again, this is a WIP. But I was expecting that the agent would update the EtcdBackup.Status, 

> User Flow
> Backup
> 5. The agent patches the EtcdBackup resource’s status to include the URL for the backup.

Reply:
Removed for now.

Reply:
I'm still unconvinced you need a manager/controller at all to do that. The controller will want to launch a Goroutine to reconcile all the EtcdBackup resources on the system. Unless we constrain it to just one. Even if you did you'd then have to somehow wait for the controller to finish reconciliation before the agent can exit, because the agent only lives for as long as the backup takes before exiting.

It sounds much easier to just use the Kubernetes client-go directly, pass in the name of the EtcdBackup resource we're targeting from the controller using a flag, and then use client-go to just issue a single update against it's status.

Reply:
> I'm still unconvinced you need a manager/controller at all to do that. 

My thinking is that the backup process has multiple points of failure, connect to etcd API, snapshot request, store to local FS, connect to backup proxy, backup request.
After each of these steps, I imagined that the Reconcile would update the EtcdBackup.Status.Phase (just as it does now)
And after each of these, it would exit the Reconcile, ready to be triggered again.
And maybe create an Event so that a user can monitor the backup progress.

> The controller will want to launch a Goroutine to reconcile all the EtcdBackup resources on the system. Unless we constrain it to just one. 

Filtering the Reconcile events to only watch a single EtcdBackup shouldn't be a problem.

> Even if you did you'd then have to somehow wait for the controller to finish reconciliation before the agent can exit, because the agent only lives for as long as the backup takes before exiting.

The agent will exit when the cluster controller deletes the backup agent pod, which it will do when it observes the EtcdBackup.Status.Phase is complete or failed.

I think it's neat to use the Reconcile pattern for the backup agent and it should allow us to re-use much of the existing controller code.

The alternative is to make the backup agent  a one-shot process and make the operator re-create the backup agent pod in the event of it failing....I'll try that first and we can then discuss it further.

> It sounds much easier to just use the Kubernetes client-go directly, pass in the name of the EtcdBackup resource we're targeting from the controller using a flag, and then use client-go to just issue a single update against it's status.

Reply:
> And after each of these, it would exit the Reconcile, ready to be triggered again.

What would trigger that Reconcile function to run again?

> The alternative is to make the backup agent a one-shot process and make the operator re-create the backup agent pod in the event of it failing....I'll try that first and we can then discuss it further.

I personally think this makes most sense - you wouldn't want a failing backup to continue to loop, potentially impacting the state of the etcd cluster.

---

Personally, I'd lean on the side of keeping it simple for now. There's nothing to say we cannot retry even if we *don't* use an informer, and I'm not sure if cramming all retrying related actions into the controller pattern brings much benefit. If we want something like that, we could create an error type `RetriableError`, use `errors.Is(RetriableError)` at the top level, and then retry there? We can then keep count of the number of failures and perma fail after some fixed number.

That said, I don't know if we *really* need retrying in the backup to be perfectly honest.. I think the cases you mention there are few and far between unless the cluster is _actually_ in a terminal/broken state (i.e. missing EtcdCluster resource or loss of quorum). I think we'd be better to wait until we run into that and then address it with a bit more knowledge/experience of the system at the time.